quint/src/quintParserFrontend.ts

```diff
@@ -114,97 +114,91 @@ export function parsePhase1(text: string, sourceLocation: string): ParseResult<P
  * Phase 2 of the Quint parser. Read the IR and check that all names are defined.
  * Note that the IR may be ill-typed.
  */
-export function parsePhase2(phase1Data: ParserPhase1):
-  ParseResult<ParserPhase2> {
-  const quintModule: QuintModule = phase1Data.module
+export function parsePhase2(phase1Data: ParserPhase1): ParseResult<ParserPhase2> {
   const sourceMap: Map<bigint, Loc> = phase1Data.sourceMap
-  const scopeTree = treeFromModule(quintModule)
-  const moduleDefinitions = collectDefinitions(quintModule)
-  const importResolvingResult = resolveImports(quintModule, moduleDefinitions)
-  let definitions: LookupTableByModule
-  const errorMessages: ErrorMessage[] = []
 
-  if (importResolvingResult.kind === 'ok') {
-    definitions = importResolvingResult.definitions
-  } else {
-    definitions = moduleDefinitions
-    importResolvingResult.errors.forEach(error => {
-      const sourceLoc = sourceMap.get(error.reference)
-      if (!sourceLoc) {
-        console.error(`No source location found for ${error.reference}. Please report a bug.`)
-      }
-      const loc = sourceLoc ?? unknownLoc
-      if (error.defName) {
-        const e =
-          `Failed to import definition ${error.defName} from module ${error.moduleName}`
-        errorMessages.push({
-          explanation: e,
-          locs: [loc],
-        })
-      } else {
-        errorMessages.push({
-          explanation: `Failed to import module ${error.moduleName}`,
-          locs: [loc],
-        })
-      }
-    })
-  }
+  return phase1Data.modules.reduce((result: ParseResult<ParserPhase2>, module) => {
+    const scopeTree = treeFromModule(module)
+    let definitions = collectDefinitions(module)
+    result.map(r => definitions = new Map([...r.table.entries(), ...definitions.entries()]))
```

Comment:
I guess it's not introduced in this pr, the would be nice to do away with the mutable `definitions` that doesn't seem necessary afaict, but it does introduce a fair bit the complexity in trying to understand this function. Perhaps something for a follow up.

Reply:
I agree. That would make this diff significantly bigger, so it would fit better as a follow up. This is part of the first chunk of code me and Igor first wrote in typescript, and it deserves some care now that we have learned how to deal with situations. I have already refactor it to use `Either` (finally!), but it's still not as nice as our newest code.